src/frontend/flight/sync.rs

```diff
@@ -371,42 +400,170 @@ impl SeafowlDataSyncManager {
+        entry: &DataSyncCollection,
+    ) -> Result<Expr> {
+        // Initialize the min/max accumulators needed to prune the table files
+        let (mut min_values, mut max_values): (Vec<MinAccumulator>, Vec<MaxAccumulator>) =
+            entry
+                .pk_columns
+                .iter()
+                .map(|pk_col| {
+                    // The validation in the interface should have ensured these fields exist.
+                    let f = full_schema.column_with_name(pk_col).unwrap().1;
+                    Ok((
+                        MinAccumulator::try_new(f.data_type())?,
+                        MaxAccumulator::try_new(f.data_type())?,
+                    ))
+                })
+                .collect::<Result<Vec<_>>>()?
+                .into_iter()
+                .unzip();
+
+        // Collect all min/max stats for PK columns
+        for sync in &entry.syncs {
+            for batch in &sync.batches {
+                entry
+                    .pk_columns
+                    .iter()
+                    .zip(min_values.iter_mut())
+                    .zip(max_values.iter_mut())
+                    .try_for_each(|((pk_col, min_value), max_value)| {
+                        let pk_array = batch.column_by_name(pk_col).unwrap().clone();
+                        min_value.update_batch(&[pk_array.clone()])?;
+                        max_value.update_batch(&[pk_array])?;
+                        Ok::<(), DataFusionError>(())
+                    })?
+            }
+        }
+
+        // Combine the statistics into a single qualifier expression
+        Ok(entry
+            .pk_columns
+            .iter()
+            .zip(min_values.iter_mut())
+            .zip(max_values.iter_mut())
+            .map(|((pk_col, min_value), max_value)| {
+                Ok(lit(pk_col)
+                    .between(lit(min_value.evaluate()?), lit(max_value.evaluate()?)))
+            })
+            .collect::<Result<Vec<Expr>>>()?
+            .into_iter()
+            .reduce(|e1: Expr, e2| e1.and(e2))
+            .unwrap())
+    }
+
+    fn apply_sync(
+        &self,
+        pk_columns: &[String],
+        full_schema: SchemaRef,
+        input_df: DataFrame,
         data: Vec<RecordBatch>,
-    ) -> Result<Arc<dyn ExecutionPlan>> {
+    ) -> Result<DataFrame> {
         let schema = data.first().unwrap().schema();
-        let mem_plan: Arc<dyn ExecutionPlan> =
-            Arc::new(MemoryExec::try_new(&[data], schema.clone(), None)?);
 
-        // TODO: Filter away deletes for now
-        let filter_plan: Arc<dyn ExecutionPlan> = Arc::new(FilterExec::try_new(
-            col(SEAFOWL_SYNC_DATA_UD_FLAG, schema.as_ref())?,
-            mem_plan,
-        )?);
+        let sync_df = self.context.inner.read_batches(data)?;
+
+        // First remove all deleted rows
+        let join_cols = pk_columns.iter().map(|pk| pk.as_str()).collect::<Vec<_>>();
+        let input_df = input_df.join(
```

Comment:
What if a row is updated multiple times in a single batch (or updated then deleted, or deleted then updated)? Will we take the latest version in that case? (I think this might need a window function with `row_number() OVER (PARTITION BY pk ORDER BY seq_number ASC`)

Reply:
Good point, haven't tried that out (for some reason i assumed this won't happen), need to think about it.